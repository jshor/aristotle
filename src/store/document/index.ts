import { StoreGetters, defineStore } from 'pinia'
import { DocumentState, state } from './state'
import * as clipboardActions from './actions/clipboard'
import * as connectionActions from './actions/connection'
import * as editorActions from './actions/editor'
import * as groupingActions from './actions/grouping'
import * as itemActions from './actions/item'
import * as panningActions from './actions/panning'
import * as portActions from './actions/port'
import * as positioningActions from './actions/positioning'
import * as oscillatorActions from './actions/oscillator'
import * as rotationActions from './actions/rotation'
import * as selectionActions from './actions/selection'
import * as simulationActions from './actions/simulation'
import * as sizingActions from './actions/sizing'
import * as snappingActions from './actions/snapping'
import * as undoRedoActions from './actions/undoRedo'
import * as zIndexActions from './actions/zIndex'
import * as zoomingActions from './actions/zooming'
import * as getters from './getters'
import sortByZIndex from '@/utils/sortByZIndex'
import BaseItem from '@/types/interfaces/BaseItem'
import Item from '@/types/interfaces/Item'
import Connection from '@/types/interfaces/Connection'

export const createDocumentStore = (documentId: string) => defineStore({
  id: documentId,
  state,
  getters: {
    ...getters,

    baseItems (state) {
      return (Object
        .values(state.connections) as BaseItem[])
        .concat(Object.values(state.items) as BaseItem[])
        .sort(sortByZIndex)
    },

    hasSelection (state) {
      return state.selectedItemIds.size > 0 || state.selectedConnectionIds.size > 0
    },

    hasSelectedItems (state) {
      return state.selectedItemIds.size > 0
    },

    canGroup (state) {
      let ungroupedCount = 0
      let groupedCount = 0

      const hasUngrouped = (list: Set<string>, map: Record<string, Connection | Item>) => {
        for (const id of list) {
          map[id].groupId
            ? groupedCount++
            : ungroupedCount++

          if (ungroupedCount > 1 || (groupedCount && ungroupedCount)) {
            // at least two items are ungrouped, or at least one item is grouped and at least one item is ungrouped
            return true
          }
        }

        return false
      }

      return hasUngrouped(state.selectedItemIds, state.items) || hasUngrouped(state.selectedConnectionIds, state.connections)
    },

    canUngroup (state) {
      return state.selectedGroupIds.size > 0
    },

    canUndo (state) {
      return state.undoStack.length > 0
    },

    canRedo (state) {
      return state.redoStack.length > 0
    }
  },
  actions: {
    ...clipboardActions,
    ...connectionActions,
    ...editorActions,
    ...groupingActions,
    ...itemActions,
    ...panningActions,
    ...portActions,
    ...positioningActions,
    ...oscillatorActions,
    ...rotationActions,
    ...selectionActions,
    ...simulationActions,
    ...snappingActions,
    ...sizingActions,
    ...undoRedoActions,
    ...zIndexActions,
    ...zoomingActions
  }
})

export type DocumentStore = ReturnType<typeof createDocumentStore>

export type DocumentStoreInstance = ReturnType<DocumentStore>
