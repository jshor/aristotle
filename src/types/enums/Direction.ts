enum Direction {
  Left = 0,
  Top,
  Right,
  Bottom
}

export default Direction
