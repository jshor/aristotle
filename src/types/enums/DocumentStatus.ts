export enum DocumentStatus {
  Ready,
  Saving,
  Printing,
  SavingImage,
  ReadOnly
}
