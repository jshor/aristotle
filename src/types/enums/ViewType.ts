export enum ViewType {
  None,
  Document,
  Preferences
}
