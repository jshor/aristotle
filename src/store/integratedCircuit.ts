import { defineStore } from 'pinia'
import { v4 as uuid } from 'uuid'
import integratedCircuitFactory from '@/factories/integratedCircuitFactory'
import FileService from '@/services/FileService'
import idMapper from '@/utils/idMapper'
import Port from '@/types/interfaces/Port'
import SerializableState from '@/types/interfaces/SerializableState'
import ItemFactory from '@/types/types/ItemFactory'
import Item from '@/types/interfaces/Item'
import Direction from '@/types/enums/Direction'
import dFlipFlop from '@/containers/fixtures/d-flipflop.json'
import tFlipFlop from '@/containers/fixtures/t-flipflop.json'
import customCircuitFactory from '@/factories/customCircuitFactory'

type IntegratedCircuitState = {
  isBuilderOpen: boolean
  model: Item | null
  factories: {
    [id: string]: ItemFactory
  }
  resolve: () => void
}

export const useIntegratedCircuitStore = defineStore('integratedCircuit', {
  state: (): IntegratedCircuitState => ({
    isBuilderOpen: false,
    model: null,
    factories: {},
    resolve: () => {}
  }),

  actions: {
    loadAllToolbox () {
      this.factories[dFlipFlop.defaultName] = customCircuitFactory.bind(null, dFlipFlop as any)
      this.factories[tFlipFlop.defaultName] = customCircuitFactory.bind(null, tFlipFlop as any)
    },

    async launchBuilder (circuit: SerializableState) {
      this.isBuilderOpen = true

      const { item } = integratedCircuitFactory(circuit, 'My integrated circuit')

      this.model = item
    },

    async loadLibrary () {
      const filePaths = window.api.getFilePaths(window.api.getDefaultSavePath(), 'aicx')

      for (let i = 0; i < filePaths.length; i++) {
        await this.loadCircuit(filePaths[i])
      }
    },

    async loadCircuit (filePath: string) {
      try {
        const content = await FileService.open(filePath)
        const parsed = JSON.parse(content) as Item

        this.factories[parsed.defaultName] = () => {
          const item = idMapper.mapIntegratedCircuitIds(parsed)
          const ports = item
            .portIds
            .reduce((map, portId) => {
              const port = item.integratedCircuit?.ports[portId]

              if (port) {
                map[port.orientation].push(port)
              }

              return map
            }, {
              [Direction.Left]: [],
              [Direction.Top]: [],
              [Direction.Right]: [],
              [Direction.Bottom]: [],
            } as Record<Direction, Port[]>)

          return { item, ports }
        }
      } catch (error) {
        console.log(error)
        window.api.showMessageBox({
          message: `Could not load ${filePath} into the integrated circuit library. The file is not a valid Aristotle circuit.`,
          title: 'Error',
          type: 'error'
        })
      }
    },

    async save () {
      if (!this.model) return

      try {
        const fileName = `${this.model.defaultName}.aicx`
        const filePath = window.api.showSaveFileDialog([
          {
            name: 'Aristotle Integrated Circuit (*.aicx)',
            extensions: ['aicx']
          }
        ], fileName)

        if (!filePath) return

        await FileService.save(filePath, this.model)
        await this.loadCircuit(filePath)

        window.api.showMessageBox({
          message: 'Integrated circuit successfully created. You can find it in the toolbox.',
          title: 'Success',
          type: 'info'
        })
      } catch (error) {
        window.api.showMessageBox({
          message: 'An error occurred while trying to save the integrated circuit.',
          title: 'Error',
          type: 'error'
        })
      }

      this.close()
    },

    close () {
      this.resolve()
      this.isBuilderOpen = true
    }
  }
})
