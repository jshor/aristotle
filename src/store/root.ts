// @ts-check
import { defineStore } from 'pinia'
import { v4 as uuid } from 'uuid'
import basic from '../containers/fixtures/basic.json'
// import flipFlop from '../containers/fixtures/flipflop.json'
import flipFlop from '../containers/fixtures/counter.json'
// import integratedCircuit from '../containers/fixtures/ic.json'
import testIc from '../containers/fixtures/test.json'

import { createDocumentStore, DocumentStore } from './document'
import { useIntegratedCircuitStore } from './integratedCircuit'
import FileService from '@/services/FileService'
import getFileName from '@/utils/getFileName'
import idMapper from '@/utils/idMapper'
import Oscillogram from '@/types/types/Oscillogram'
import Item from '@/types/interfaces/Item'
import { DocumentStatus } from '@/types/enums/DocumentStatus'
import { ViewType } from '@/types/enums/ViewType'

export type RootStore = {
  documents: {
    [id: string]: {
      fileName: string
      filePath?: string
      displayName: string
      store: DocumentStore
    }
  }
  clipboard: string | null
  activeDocumentId: string | null
  canExit: boolean
  canPaste: boolean
  isDocumentSelectOpen: boolean
  isMobilePulloutOpen: boolean
  isFullscreen: boolean
  dialogType: ViewType
  isToolboxOpen: boolean
  navigationAnimationFrameId: number
  closeDialog?: () => Promise<boolean>
}

export const useRootStore = defineStore({
  id: 'root',
  state: (): RootStore => ({
    documents: {},
    clipboard: null,
    activeDocumentId: null,
    canExit: false,
    canPaste: false,
    isFullscreen: false,
    isToolboxOpen: false,
    isMobilePulloutOpen: false,
    dialogType: ViewType.None,
    isDocumentSelectOpen: false,
    navigationAnimationFrameId: 0
  }),
  getters: {
    hasOpenDocuments (state) {
      return Object
        .keys(state.documents)
        .length > 0
    },
    activeDocument (state) {
      if (state.activeDocumentId) {
        return state.documents[state.activeDocumentId]
      }

      return null
    },
    viewType (state) {
      if (state.dialogType) {
        return state.dialogType
      }

      return state.documents[state.activeDocumentId!]
        ? ViewType.Document
        : ViewType.None
    }
  },
  actions: {
    async launchIntegratedCircuitBuilder () {
      const integratedCircuitStore = useIntegratedCircuitStore()
      const document = this.activeDocument

      if (document) {
        await integratedCircuitStore.launchBuilder(document.store().$state)

        const store = document.store()

        // const dialogResult = window.api.showMessageBox({
        //   message: `Integrated circuit successfully created. Do you want to add it to the current document?`,
        //   title: 'Confirm',
        //   buttons: ['Yes', 'No']
        // })

        const idMappedIcItem = idMapper.mapIntegratedCircuitIds(integratedCircuitStore.model as Item)

        store.insertItemAtPosition({ item: idMappedIcItem })
        store.setItemBoundingBox(idMappedIcItem.id)
        store.setItemSelectionState(idMappedIcItem.id, true)
      }
    },
    openIntegratedCircuit (item: Item) {
      if (item.integratedCircuit) {
        this.openDocument('', item.integratedCircuit.serializedState, item.id, item.defaultName)
      }
    },
    async saveImage (data: Blob) {
      if (!this.activeDocument) return

      try {
        const partialName = this.activeDocument.displayName.replace(/\.[^/.]+$/, "")
        const fileName = `${partialName}.png`
        const filePath = window.api.showSaveFileDialog([
          {
            name: 'PNG Image',
            extensions: ['.png']
          }
        ], fileName)

        if (!filePath) return

        await FileService.writeBlob(filePath, data)
      } catch (error) {
        console.log('ERROR: ', error)
        window.api.showMessageBox({
          message: 'An error occurred while trying to save the image.',
          title: 'Error',
          type: 'error'
        })
      }
    },
    newDocument () {
      const id = uuid()
      const store = createDocumentStore(id)
      const document = store()

      document.initialize()

      this.documents[id] = {
        fileName: 'Untitled Circuit',
        displayName: 'Untitled Circuit',
        store
      }
      this.activateDocument(id)
    },

    async closeApplication (): Promise<boolean> {
      const dialogCloseResult = await this.closeDialog?.()

      if (dialogCloseResult === false) {
        return false
      }

      return this.closeAllDocuments()
    },

    /**
     * Attempts to close all open documents.
     *
     * @returns true if all documents were closed successfully, false otherwise
     */
    async closeAllDocuments (): Promise<boolean> {
      const openDocuments = [this.activeDocumentId!].concat(Object.keys(this.documents))

      for (let i = 0; i < openDocuments.length; i++) {
        if (this.documents[openDocuments[i]]) {
          const closedSuccessfully = await this.closeDocument(openDocuments[i])

          if (!closedSuccessfully) {
            return false
          }
        }
      }

      return true
    },

    /**
     * Closes the currently-active document.
     */
    async closeActiveDocument () {
      if (!this.activeDocumentId) return

      const closedSuccessfully = await this.closeDocument(this.activeDocumentId)

      if (closedSuccessfully) {
        await this.switchDocument(1)
      }
    },

    /**
     * Closes the document of the given ID.
     * This will prompt the user to save the document if it has unsaved changes.
     *
     * @returns true if the document was closed successfully, false otherwise
     */
    async closeDocument (id: string): Promise<boolean> {
      const document = this.documents[id]
      const store = document.store()

      if (!store.isDirty) {
        if (this.activeDocumentId === id) {
          this.switchDocument(1)
        }
        delete this.documents[id]
        return true
      }

      this.activateDocument(id)

      await store.load()

      const saveDialogResult = window
        .api
        .showMessageBox({
          message: `Save changes to ${document.displayName}?`,
          title: 'Confirm',
          buttons: ['Yes', 'No', 'Cancel']
        })

      if (saveDialogResult === 2) return false
      if (saveDialogResult === 0) {
        try {
          if (!this.attemptDocumentSave()) {
            return false
          }
        } catch (error) {
          const errorDialogResult = window
            .api
            .showMessageBox({
              message: 'An error occurred while trying to save the file. Close anyway?',
              title: 'Error',
              type: 'error',
              buttons: ['Yes', 'No']
            })

          if (errorDialogResult === 1) {
            return false
          }
        }
      }

      store.stopSimulation()
      store.$reset()
      store.$dispose()

      delete this.documents[id]

      return true
    },

    /**
     * Saves the document.
     *
     * @param saveAsNewFile - if true, forces saving the document as a new file
     */
    saveActiveDocument (saveAsNewFile: boolean = false) {
      try {
        this.attemptDocumentSave(saveAsNewFile)
      } catch (error) {
        window
          .api
          .showMessageBox({
            message: 'An error occurred while trying to save the file.',
            title: 'Error',
            type: 'error'
          })
      }
    },

    /**
     * Saves all open documents.
     */
    saveAllDocuments () {
      throw new Error('Not implemented.')
    },

    /**
     * Attempts to save the document as a file.
     * If the document has not been saved before, it will prompt the user to select a location to save the file.
     *
     * @param saveAsNewFile - if true, forces saving the document as a new file
     */
    attemptDocumentSave (saveAsNewFile: boolean = false) {
      if (!this.activeDocument) return true
      if (!this.activeDocument.filePath || saveAsNewFile) {
        // if no file name is defined yet, ask the user for the location to save it
        const fileName = this.activeDocument.fileName || `${this.activeDocument.displayName}.alfx`
        const filePath = window.api.showSaveFileDialog([
          {
            name: 'Aristotle Logic Circuit (*.alfx)',
            extensions: ['alfx']
          }
        ], fileName)

        if (filePath) {
          const fileName = getFileName(filePath)

          this.activeDocument.filePath = filePath
          this.activeDocument.fileName = fileName
          this.activeDocument.displayName = fileName
        }

        return false
      }

      const store = this.activeDocument.store()

      FileService.save(this.activeDocument.filePath!, {
        items: store.items,
        connections: store.connections,
        groups: store.groups,
        ports: store.ports
      })

      store.isDirty = false

      return true
    },
    async selectDocument () {
      const files = window.api.showOpenFileDialog([
        {
          name: 'Aristotle Logic Circuit (*.alfx)',
          extensions: ['alfx']
        },
        {
          name: 'Aristotle Integrated Circuit (*.aicx)',
          extensions: ['aicx']
        }
      ])

      for (let i = 0; i < files.length; i++) {
        await this.openDocumentFromPath(files[i])
      }
    },
    async openDocumentFromPath (filePath: string) {
      try {
        // TODO: need to create lock on file while open
        const content = await FileService.open(filePath)

        if (content) {
          this.openDocument(filePath, content, uuid())
        }
      } catch (error) {
        console.log(error)
        window.api.showMessageBox({
          message: `Could not open ${filePath}. The file is not a valid Aristotle circuit.`,
          title: 'Error',
          type: 'error'
        })
      }
    },
    openDocument (filePath: string, content: string, id: string, displayName: string = '') {
      if (this.documents[id]) {
        return this.activateDocument(id)
      }

      const store = createDocumentStore(id)
      const document = store()
      const fileName = getFileName(filePath)

      document.loadDocument(content)
      document.oscillator.onTick = (o: Oscillogram) => document.oscillogram = o

      this.documents[id] = {
        fileName,
        // if this is an integrated circuit file (.aicx), then it cannot be directly saved
        filePath: /^.*.aicx$/i.test(filePath)
          ? undefined // omit the file path for IC files
          : filePath, // otherwise, use the file path for a typical circuit
        displayName: displayName || fileName,
        store
      }
      this.activateDocument(id)
    },
    pauseActivity () {
      this.activeDocument?.store().stopSimulation()
    },
    resumeActivity () {
      this.activeDocument?.store().startSimulation()
    },
    activateDocument (id: string) {
      this.pauseActivity()
      this.activeDocumentId = id
      this.resumeActivity()
    },
    openTestDocuments () {
      this.openDocument('flip-flop.alfx', JSON.stringify(flipFlop), uuid(), 'flip-flop.alfx')
    },

    switchDocument (direction: number) {
      if (this.navigationAnimationFrameId) return

      this.navigationAnimationFrameId = requestAnimationFrame(async () => {
        const documentIds = Object.keys(this.documents)
        const currentIndex = documentIds.findIndex(id => id === this.activeDocumentId)
        let nextIndex = currentIndex + direction

        if (nextIndex < 0) nextIndex = documentIds.length - 1
        if (nextIndex >= documentIds.length) nextIndex = 0

        this.activateDocument(documentIds[nextIndex])
        await this.activeDocument?.store().load()
        this.navigationAnimationFrameId = 0
      })
    },

    toggleFullscreen () {
      this.isFullscreen = !this.isFullscreen

      window.api.setFullscreen(this.isFullscreen)
    },

    toggleToolbox () {
      this.isToolboxOpen = !this.isToolboxOpen
    },

    checkPasteability () {
      this.canPaste = window.api.canPaste()
    },

    setAllDocumentStatuses (status: DocumentStatus) {
      Object
        .keys(this.documents)
        .forEach(id => this.documents[id].store().setStatus(status))
    }
  }
})
