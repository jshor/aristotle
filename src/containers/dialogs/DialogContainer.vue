<template>
  <modal
    v-if="!isClosing"
    v-model="isOpen"
    :title="title"
  >
    <slot />
  </modal>
</template>

<script lang="ts" setup>
import { computed, watch, ref } from 'vue'
import Modal from '@/components/modal/Modal.vue'
import { ViewType } from '@/types/enums/ViewType'
import { useRootStore } from '@/store/root'

const rootStore = useRootStore()
const props = defineProps<{
  dialogType: ViewType
  title: string
  onBeforeClose?: () => Promise<boolean>
}>()
const isClosing = ref(false)
const isOpen = computed({
  get () {
    return rootStore.dialogType === props.dialogType
  },
  set (value) {
    if (!value) {
      closeDialog(false)
    }
  }
})

/**
 * Close the dialog.
 */
async function closeDialog (immediate = true) {
  let result = true

  if (props.onBeforeClose) {
    result = await props.onBeforeClose()
  }

  isClosing.value = immediate
  rootStore.dialogType = ViewType.None

  await new Promise(resolve => setTimeout(resolve))

  isClosing.value = false

  return result
}

watch(isOpen, value => {
  rootStore.closeDialog = value
    ? closeDialog
    : undefined
})
</script>
