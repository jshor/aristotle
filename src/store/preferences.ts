import Property from '@/types/interfaces/Property'
import { defineStore } from 'pinia'
import { useRootStore } from './root'

export const usePreferencesStore = defineStore({
  id: 'preferences',
  state: () => ({
    isOpen: false,
    colors: {
      onColor: {
        type: 'color',
        label: 'High state',
        value: '#4bc567',
        description: 'When a signal is present.'
      } as Property<string>,
      offColor: {
        type: 'color',
        label: 'Low state',
        value: '#c0c0c0',
        description: 'When no signal is present.'
      } as Property<string>,
      unknownColor: {
        type: 'color',
        label: 'Unknown state',
        value: '#ff0000',
        description: 'High impedance (Hi-Z).'
      } as Property<string>,
      animate: {
        type: 'boolean',
        label: 'Animate signal direction',
        value: true,
        description: 'Show animated dashes in the direction of the signal\'s electrical flow.'
      } as Property<boolean>
    },
    snapping: {
      snapToGrid: {
        type: 'boolean',
        label: 'Snap to grid',
        value: false,
        description: 'Snap elements to the grid lines.',
        excludes: ['snapToAlign']
      } as Property<boolean>,
      snapToAlign: {
        type: 'boolean',
        label: 'Snap to align',
        value: true,
        description: 'Snap elements together.',
        excludes: ['snapToGrid']
      } as Property<boolean>,
      /** The distance to the element to snap to. */
      snapTolerance: {
        type: 'number',
        label: 'Snap tolerance',
        /** The tolerance value. */
        value: 15,
        min: 1,
        max: 50,
        description: 'Maximum distance from a dragging object to snap to another element.'
      } as Property<number>
    },
    grid: {
      showGrid: {
        type: 'boolean',
        label: 'Show grid',
        value: true,
        description: 'Display a grid on the canvas.'
      } as Property<boolean>,
      /** The grid size to display. */
      gridSize: {
        type: 'number',
        label: 'Grid size',
        /** The width/height of each grid cell. */
        value: 20,
        min: 10,
        max: 50,
        description: 'Distance between grid lines.',
        excludes: ['snapToAlign']
      } as Property<number>
    },
    simulation: {
      startPaused: {
        type: 'boolean',
        label: 'Pause on startup',
        value: true,
        description: 'Start simulation paused.'
      } as Property<boolean>
    },
    experience: {
      vibration: {
        type: 'boolean',
        label: 'Vibration',
        value: true,
        description: 'Haptic feedback on gestures.',
        mobileOnly: true
      } as Property<boolean>,
      autosave: {
        type: 'boolean',
        label: 'Auto-save',
        value: true,
        description: 'Automatically save your work.',
        desktopOnly: true
      } as Property<boolean>,
      darkMode: {
        type: 'boolean',
        label: 'Dark mode',
        value: true
      } as Property<boolean>
    }
  }),

  getters: {
    /**
     * The CSS color variables.
     */
    colorStyles (state) {
      return {
        '--color-on': state.colors.onColor.value,
        '--color-off': state.colors.offColor.value,
        '--color-hi-z': state.colors.unknownColor.value
      }
    }
  },

  actions: {
    async savePreferences () {
      try {
        // TODO: determine settings path for each OS
        // await FileService.open('?')
      } catch (_) {}
    }
  }
})
