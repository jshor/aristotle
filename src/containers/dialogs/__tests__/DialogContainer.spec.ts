import { shallowMount, VueWrapper } from '@vue/test-utils'
import { setActivePinia, createPinia } from 'pinia'
import DialogContainer from '../DialogContainer.vue'
import { useRootStore } from '@/store/root'
import { ViewType } from '@/types/enums/ViewType'

setActivePinia(createPinia())

describe('Dialog Container', () => {
  let wrapper: VueWrapper<typeof DialogContainer>
  let store: ReturnType<typeof useRootStore>

  const title = 'Dialog Test'
  const dialogType = ViewType.Preferences

  beforeEach(() => {
    store = useRootStore()
    wrapper = shallowMount(DialogContainer, {
      props: {
        title,
        dialogType
      }
    })
  })

  afterEach(() => {
    jest.resetAllMocks()
    store.$reset()
  })

  it('should match the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  describe('when the dialog is opened', () => {
    beforeEach(() => {
      store.$patch({
        dialogType
      })
    })

    it('should open the modal', () => {
      expect(wrapper.findComponent({ name: 'Modal' }).props()).toEqual(expect.objectContaining({
        modelValue: true
      }))
    })

    it('should close the modal when the dialog type changes', async () => {
      store.$patch({
        dialogType: ViewType.None
      })
      await wrapper.vm.$nextTick()
      expect(wrapper.findComponent({ name: 'Modal' }).props()).toEqual(expect.objectContaining({
        modelValue: false
      }))

      store.$patch({
        dialogType: ViewType.Preferences
      })
      await wrapper.vm.$nextTick()
      expect(wrapper.findComponent({ name: 'Modal' }).props()).toEqual(expect.objectContaining({
        modelValue: true
      }))
    })

    it('should close the modal when the modal itself requests it', async () => {
      const modal = wrapper.findComponent({ name: 'Modal' })

      await modal.vm.$emit('update:modelValue', false)
      await wrapper.vm.$nextTick()

      expect(modal.props()).toEqual(expect.objectContaining({
        modelValue: false
      }))
    })

    it('should kill the modal element when the root store closes it', async () => {
      store.closeDialog!()

      await wrapper.vm.$nextTick()

      expect(wrapper.findComponent({ name: 'Modal' }).exists()).toBe(false)
    })
  })

  describe('when the dialog is closing', () => {
    it('should invoke onBeforeClose() before closing the dialog', async () => {
      store.$patch({
        dialogType: ViewType.Preferences
      })

      const onBeforeClose = jest.fn()
      await wrapper.setProps({
        title,
        dialogType,
        onBeforeClose
      })
      const modal = wrapper.findComponent({ name: 'Modal' })
      await modal.vm.$emit('update:modelValue', false)

      expect(onBeforeClose).toHaveBeenCalledTimes(1)
    })
  })

  describe('when the dialog is closed', () => {
    it('should not open the dialog even if the modal itself requests it', async () => {
      store.$patch({
        dialogType: ViewType.None
      })

      const modal = wrapper.findComponent({ name: 'Modal' })

      await modal.vm.$emit('update:modelValue', true)
      await wrapper.vm.$nextTick()

      expect(modal.props()).toEqual(expect.objectContaining({
        modelValue: false
      }))
    })
  })
})
