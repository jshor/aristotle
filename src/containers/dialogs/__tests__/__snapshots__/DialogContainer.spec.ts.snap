// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Dialog Container > should match the snapshot 1`] = `"<modal-stub modelvalue="false" title="Dialog Test"></modal-stub>"`;

exports[`Dialog Container should match the snapshot 1`] = `"<modal-stub modelvalue="false" title="Dialog Test"></modal-stub>"`;

exports[`Toolbox Item should match the snapshot 1`] = `"<modal-stub modelvalue="false" title="Dialog Test"></modal-stub>"`;
